modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/ComponentMessageProcessor.java

```diff
@@ -437,18 +441,62 @@ private void onEventSynchronous(CoreEvent event, ExecutorCallback executorCallba
         prepareAndExecuteOperation(event, () -> callback, ctx);
       };
 
-      operationExecutionFunction.execute(resolutionResult, event, executorCallback);
+      operationExecutionFunction.execute(resolutionResult, event,
+                                         computeOperationReturnExecutionCallback(event, executorCallback));
     } catch (Throwable t) {
       executorCallback.error(unwrap(t));
     }
   }
 
+  private ExecutorCallback computeOperationReturnExecutionCallback(CoreEvent event, ExecutorCallback delegateCallback) {
+    return new ExecutorCallback() {
+
+      @Override
+      public void complete(Object o) {
+        ExecutionContextAdapter operationContext = null;
+        CoreEvent originalEvent = null;
+        try {
+          if (returnDelegate instanceof ValueReturnDelegate) {
```

Comment:
I would make a best effort to avoid this. This only makes sense in the particular policy use case you're targeting. This is wrong in any other use case. So either do this differently, or make it absolutely impossible for this method to be used in any other use case.